config/confighttp/compressor.go

```diff
@@ -23,45 +24,72 @@ type writeCloserReset interface {
 	Reset(w io.Writer)
 }
 
-var (
-	_          writeCloserReset = (*gzip.Writer)(nil)
-	gZipPool                    = &compressor{pool: sync.Pool{New: func() any { return gzip.NewWriter(nil) }}}
-	_          writeCloserReset = (*snappy.Writer)(nil)
-	snappyPool                  = &compressor{pool: sync.Pool{New: func() any { return snappy.NewBufferedWriter(nil) }}}
-	_          writeCloserReset = (*zstd.Encoder)(nil)
-	// Concurrency 1 disables async decoding via goroutines. This is useful to reduce memory usage and isn't a bottleneck for compression using sync.Pool.
-	zStdPool                  = &compressor{pool: sync.Pool{New: func() any { zw, _ := zstd.NewWriter(nil, zstd.WithEncoderConcurrency(1)); return zw }}}
-	_        writeCloserReset = (*zlib.Writer)(nil)
-	zLibPool                  = &compressor{pool: sync.Pool{New: func() any { return zlib.NewWriter(nil) }}}
-	_        writeCloserReset = (*lz4.Writer)(nil)
-	lz4Pool                   = &compressor{pool: sync.Pool{New: func() any {
-		lz := lz4.NewWriter(nil)
-		// Setting concurrency to 1 to disable async decoding by goroutines. This will reduce the overall memory footprint and pool
-		_ = lz.Apply(lz4.ConcurrencyOption(1))
-		return lz
-	}}}
-)
+type compressorMap struct {
+	pools map[string]*compressor
+}
 
 type compressor struct {
 	pool sync.Pool
 }
 
+var (
+	compressorPools                   = &compressorMap{pools: make(map[string]*compressor)}
+	snappyCompressor                  = &compressor{}
+	lz4Compressor                     = &compressor{}
+	_                writeCloserReset = (*gzip.Writer)(nil)
+	_                writeCloserReset = (*snappy.Writer)(nil)
+	_                writeCloserReset = (*zstd.Encoder)(nil)
+	_                writeCloserReset = (*zlib.Writer)(nil)
+	_                writeCloserReset = (*lz4.Writer)(nil)
+)
+
 // writerFactory defines writer field in CompressRoundTripper.
 // The validity of input is already checked when NewCompressRoundTripper was called in confighttp,
-func newCompressor(compressionType configcompression.Type) (*compressor, error) {
+func newCompressor(compressionType configcompression.Type, compressionParams configcompression.CompressionParams) (*compressor, error) {
```

Comment:
This function is no longer concurrent-safe, and needs to be. Accessing/updating `compressorPools` should be guarded by a mutex.